src/headingcommand.js

```diff
@@ -20,176 +21,116 @@ export default class HeadingCommand extends Command {
 	 * @param {module:engine/model/batch~Batch} [options.batch] Batch to collect all the change steps.
 	 * New batch will be created if this option is not set.
 	 */
 	_doExecute( options = {} ) {
-		// TODO: What should happen if option is not found?
-		const id = options.id || this.defaultOption.id;
-		const doc = this.editor.document;
-		const selection = doc.selection;
-		const startPosition = selection.getFirstPosition();
-		const elements = [];
-		// Storing selection ranges and direction to fix selection after renaming. See ckeditor5-engine#367.
+		const editor = this.editor;
+		const document = editor.document;
+		const selection = document.selection;
 		const ranges = [ ...selection.getRanges() ];
 		const isSelectionBackward = selection.isBackward;
-		// If current option is same as new option - toggle already applied option back to default one.
-		const shouldRemove = ( id === this.value.id );
-
-		// Collect elements to change option.
-		// This implementation may not be future proof but it's satisfactory at this stage.
-		if ( selection.isCollapsed ) {
-			const block = findTopmostBlock( startPosition );
 
-			if ( block ) {
-				elements.push( block );
-			}
-		} else {
-			for ( let range of ranges ) {
-				let startBlock = findTopmostBlock( range.start );
-				const endBlock = findTopmostBlock( range.end, false );
-
-				elements.push( startBlock );
-
-				while ( startBlock !== endBlock ) {
-					startBlock = startBlock.nextSibling;
-					elements.push( startBlock );
-				}
-			}
-		}
+		// If current option is same as new option - toggle already applied option back to default one.
+		const shouldRemove = this.value;
 
-		doc.enqueueChanges( () => {
-			const batch = options.batch || doc.batch();
+		document.enqueueChanges( () => {
+			const batch = options.batch || document.batch();
 
-			for ( let element of elements ) {
+			for ( let element of document.selection.getSelectedBlocks() ) {
 				// When removing applied option.
 				if ( shouldRemove ) {
-					if ( element.name === id ) {
-						batch.rename( element, this.defaultOption.id );
+					if ( element.name === this.modelElement ) {
+						// Apply paragraph to the selection withing that particular element only instead
+						// of working on the entire document selection.
+						const selection = new Selection();
+						selection.addRange( Range.createIn( element ) );
+
+						// Share the batch with the paragraph command.
+						editor.execute( 'paragraph', { selection, batch } );
 					}
 				}
 				// When applying new option.
 				else {
-					batch.rename( element, id );
+					batch.rename( element, this.modelElement );
 				}
 			}
 
 			// If range's selection start/end is placed directly in renamed block - we need to restore it's position
 			// after renaming, because renaming puts new element there.
-			doc.selection.setRanges( ranges, isSelectionBackward );
+			selection.setRanges( ranges, isSelectionBackward );
```

Comment:
PS. This might've been necessary in the past when renaming wasn't an operation because then the live selection  transformations might not have been able to keep the selection position.